Fix over-allocation of mounted indices on a cold/frozen node

arteam opened the pull request:
`ClusterServiceInfo` can have stale information about free space in disk because gets updated periodically. If we allocate a lot of searchable snapshots on a cold node, we can break disk thresholds and over allocate on a node where there's no disk space left.

Preserve the expected shard size in `ShardRouting` when starting a new shard in case it was recovered from a searchable snapshot. Cache them in `unaccountedSearchableSnapshotSizes` for `RoutingAllocation`, so `DiskThresholdDecider` can account for the searchable snapshot shards that haven't been synced with cluster service info yet.

Fixes https://github.com/elastic/elasticsearch/pull/85018

arteam committed:
Copy the expected shard size for STARTED shards

elasticmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

arteam committed:
Update docs/changelog/86331.yaml

elasticsearchmachine commented:
Hi @arteam, I've created a changelog YAML for you.

arteam committed:
Merge remote-tracking branch 'origin/master' into copy-expected-shard-size

arteam committed:
Merge remote-tracking branch 'origin/master' into copy-expected-shard-size

arteam committed:
Remove comment

arteam committed:
Add comment about shared cache

arteam commented:
@elasticmachine update branch

arteam committed:
Merge remote-tracking branch 'origin/master' into copy-expected-shard-size

arteam commented:
@elasticmachine update branch

arteam committed:
Merge remote-tracking branch 'origin/master' into copy-expected-shard-size

arteam committed:
Merge remote-tracking branch 'origin/master' into copy-expected-shard-size

arteam committed:
Merge remote-tracking branch 'origin/master' into copy-expected-shard-size

arteam committed:
Make sure we calculate started shards only for searchable snapshots

arteam committed:
Rename routing to shard

arteam commented:
@elasticmachine run elasticsearch-ci/part-1

arteam committed:
Don't output expected shard size for started shards in summary

arteam committed:
Don't output expected shard size for started shards in XContent

arteam committed:
Adjust ShardRoutingTests for not printing expected size

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDecider.java:
```diff
@@ -144,6 +143,24 @@ public static long sizeOfRelocatingShards(
+            if (clusterInfo.getShardSize(shard) != null) {
+                continue;
+            }
+            totalSize += Math.max(shard.getExpectedShardSize(), 0L);
```
I think we need this in the INITIALIZING case too for searchable snapshots? (or rather use the `shard.getExpectedShardSize` value in place of the default `0L` there).

I wonder if we can make a test that fail in that case. I.e., we have X initializing shards, allowing us to go X-1 shards too high.

fcofdez replied:
Yes, since we only account for relocating shards for the `INITIALIZING` shards. I agree that it would be nice to add a test for that case.

arteam replied:
I've changed `0L` to `Math.max(routing.getExpectedShardSize(), 0L)` for the `INITIALIZING` case, too, but I couldn't come up with a test case showing that it makes any difference :(

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java:
```diff
@@ -505,7 +502,7 @@ public ShardRouting moveToStarted() {
             null,
             allocationId,
-            UNAVAILABLE_EXPECTED_SHARD_SIZE
+            expectedShardSize
```
We talked about only keeping this for searchable snapshots, where the size is immutable.

arteam replied:
Done!

arteam replied:
I pass `Metatada` to `createStarted` in order to figure whether a new shard is searchable snapshot

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java:
```diff
@@ -748,7 +745,7 @@ public String shortSummary() {
             sb.append(", ").append(unassignedInfo.toString());
         }
-        if (expectedShardSize != UNAVAILABLE_EXPECTED_SHARD_SIZE) {
+        if (expectedShardSize != UNAVAILABLE_EXPECTED_SHARD_SIZE && state != ShardRoutingState.STARTED) {
```
I find this counter intuitive. If there is a shard size, I think we'd want the short summary and toString to contain it?

arteam replied:
I believe there was an integration state checking the cluster state that relied on the `toString` output and it was failing because of the changed format. I believe it's should be ok to remove the check after we made sure `expectedShardSize` is set only for searchable snapshots.

arteam replied:
Fixed it!

henningandersen reviewed (commented):
This direction looks good. I left a couple of comments to address.

I'd like to also see a test in `DiskThresholdDeciderUnitTests`, validating that the extra information is taken into account.

arteam committed:
Merge branch 'master' into copy-expected-shard-size

arteam committed:
Merge remote-tracking branch 'origin/master' into copy-expected-shard-size

arteam committed:
Use shard.getExpectedShardSize for the INITIALIZED case

arteam committed:
Store expectedShareSize only for searchable snapshots

arteam committed:
Print expectedShardSize in the text output

arteam committed:
Add comment about disabling merges

arteam committed:
Make sure we test only cold node

arteam commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/allocation/SearchableSnapshotDiskThresholdIntegTests.java:
```diff
@@ -105,6 +106,7 @@ public void testHighWatermarkCanBeExceededOnColdOrFrozenNode() throws Exception
                             .put(INDEX_SOFT_DELETES_SETTING.getKey(), true)
                             .put(INDEX_STORE_STATS_REFRESH_INTERVAL_SETTING.getKey(), "0ms")
                             .put(DataTier.TIER_PREFERENCE_SETTING.getKey(), DataTier.DATA_HOT)
+                            .put(MergePolicyConfig.INDEX_MERGE_ENABLED, false)
```
Yes, I've added a comment explaining it :+1:

arteam committed:
Merge remote-tracking branch 'origin/master' into copy-expected-shard-size

arteam committed:
Merge remote-tracking branch 'origin/master' into copy-expected-shard-size

arteam committed:
Add unit test for taking into account expected size for searchable snapshots

arteam committed:
Prevent overcomitting for searchable snapshots

arteam committed:
Merge remote-tracking branch 'origin/master' into copy-expected-shard-size

arteam committed:
Don't use ignored shard, just reduce the size in order not to fit all the shards

arteam committed:
Merge remote-tracking branch 'origin/master' into copy-expected-shard-size

arteam committed:
We only need to reduce the size on one in order not to fit one shard

arteam committed:
Make sure the server is red

arteam committed:
Extract common methods

arteam committed:
Extract storage as a parameter

arteam committed:
Check multiple searchable snapshot shards

arteam committed:
Rename to sizeOfUnaccountedShards

fcofdez commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/allocation/SearchableSnapshotDiskThresholdIntegTests.java:
```diff
@@ -268,6 +275,73 @@ public void testHighWatermarkCanBeExceededOnColdOrFrozenNode() throws Exception
+        var mockRepository = (MockRepository) internalCluster().getCurrentMasterNodeInstance(RepositoriesService.class)
+            .repository("repository");
+        // Prevent searchable snapshot shards from quickly jumping from INITIALIZED to STARTED
+        mockRepository.setBlockOnceOnReadSnapshotInfoIfAlreadyBlocked();
```
Let's move this before we mount the snapshot, to avoid possible interactions while we take the snapshot.

arteam replied:
I believe we already do. The `createSnapshot("repository", "snapshot", nbIndices)` line is after the `setBlockOnceOnReadSnapshotInfoIfAlreadyBlocked` call

fcofdez replied:
I meant to move it before we call `mountIndices(indicesStoresSizes, "mounted-", "repository", "snapshot", FULL_COPY);`

arteam replied:
Moved

fcofdez commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/allocation/SearchableSnapshotDiskThresholdIntegTests.java:
```diff
@@ -268,6 +275,73 @@ public void testHighWatermarkCanBeExceededOnColdOrFrozenNode() throws Exception
+        );
+        ensureStableCluster(3);
+
+        String otherDataNodeId = internalCluster().getInstance(NodeEnvironment.class, otherDataNode).nodeId();
```
nit: I think it's better to rely on the cluster state for this:
```
var clusterState = client().admin().cluster().prepareState().get().getState();
var nodeId = clusterState.nodes().resolveNode(otherNadaNode).getId();
```

arteam replied:
Done!

fcofdez commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/allocation/SearchableSnapshotDiskThresholdIntegTests.java:
```diff
@@ -268,6 +275,73 @@ public void testHighWatermarkCanBeExceededOnColdOrFrozenNode() throws Exception
+        InternalClusterInfoService masterInfoService = (InternalClusterInfoService) internalCluster().getCurrentMasterNodeInstance(
+            ClusterInfoService.class
+        );
+        ClusterInfoServiceUtils.refresh(masterInfoService);
```
do we need to refresh the cluster info at this point?

fcofdez replied:
But we already call just before the assertion, right?
```
        ClusterInfoServiceUtils.refresh(masterInfoService);
        assertThat(
            masterInfoService.getClusterInfo().getNodeMostAvailableDiskUsages().get(otherDataNodeId).getTotalBytes(),
            equalTo(totalSpace)
        );
```

arteam replied:
Right, we need to do `refresh` only once :+1:

fcofdez commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/allocation/SearchableSnapshotDiskThresholdIntegTests.java:
```diff
@@ -268,6 +275,73 @@ public void testHighWatermarkCanBeExceededOnColdOrFrozenNode() throws Exception
 
+    public void testOvercommitInitializingSearchableSnapshotShards() throws Exception {
+        internalCluster().startMasterOnlyNode();
+        internalCluster().startNode(
```
nit: shorter version `internalCluster().startNode(onlyRole(DATA_HOT_NODE_ROLE));`

fcofdez commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/allocation/SearchableSnapshotDiskThresholdIntegTests.java:
```diff
@@ -155,11 +147,67 @@ public void testHighWatermarkCanBeExceededOnColdOrFrozenNode() throws Exception
+        mountLatch.await();
+    }
+
+    public void testHighWatermarkCanBeExceededOnColdNode() throws Exception {
```
s/testHighWatermarkCanBeExceededOnColdNode/testHighWatermarkCanNotBeExceededOnColdNode/ :)

fcofdez commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/allocation/SearchableSnapshotDiskThresholdIntegTests.java:
```diff
@@ -268,6 +275,73 @@ public void testHighWatermarkCanBeExceededOnColdOrFrozenNode() throws Exception
+            equalTo(totalSpace)
+        );
+
+        mountIndices(indicesStoresSizes, "mounted-", "repository", "snapshot", FULL_COPY);
```
We should validate that we do not allow allocating more shards that go over the disk threshold, right?

arteam replied:
Exactly! We check that we can't allocate N shards when there's a space only for N-1 shards

fcofdez replied:
Can we change the test to show that we indeed allocate `N-1` shards and the extra shards is not allocated?

fcofdez commented on server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java:
```diff
@@ -484,18 +483,27 @@ public ShardRouting reinitializeReplicaShard() {
      * canceled.
      */
-    public ShardRouting moveToStarted() {
+    public ShardRouting moveToStarted(@Nullable Metadata metadata) {
```
I think we should have an extra field within `ShardRouting` to tell whether or not this shard is a searchable snapshot, that should simplify things and allows us to have clearer assertions. wdyt?

fcofdez replied:
I think it would add clarity and simplify this where we now have more API surface, i.e. why do we have 2 `moveToStarted` methods? additionally we could add more assertions to ensure that we only populate the expected shard size when it's needed.

arteam replied:
WDYT about addressing this in a separate PR? `ShardRouting`'s constructor gets called from a lot of places, so adding a new field would add a significant amount of changes in the classes unrelated to the PR.

henningandersen replied:
I would also like to get rid of the extra `moveToStarted` variant. Being lenient and assuming null in this central place seems dangerous.

But I do see the problems in doing so, at least in the test code, where we often do not have `Metadata` nor `IndexMetadata`. I wonder if we could fix this by instead letting `moveToStarted` take `expectedShardSize` as parameter. We can then let `AllocationService derive the `expectedShardSize` from the metadata it already has available.

I think this would help separate responsibilities too.

All tests that do not care can then pass in `UNAVAILABLE_EXPECTED_SHARD_SIZE`.

fcofdez reviewed (commented):
Thanks for adding the extra test, the direction looks good. I left a few extra comments.

arteam committed:
Merge remote-tracking branch 'origin/master' into copy-expected-shard-size

arteam committed:
Unblock repository

arteam commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/allocation/SearchableSnapshotDiskThresholdIntegTests.java:
```diff
@@ -268,6 +275,73 @@ public void testHighWatermarkCanBeExceededOnColdOrFrozenNode() throws Exception
+                ClusterHealthStatus.RED,
+                client().admin().cluster().health(new ClusterHealthRequest()).actionGet().getStatus()
+            )
+        );
```
Added `mockRepository.unblock()` before checking the cluster health status. Did you men that?

fcofdez replied:
My idea was:
1. Unblock the repo
2. Check that expected shards move to started

I feel like asserting that the cluster is RED is a bit coarse.

arteam committed:
Get otherNodeId from the cluster state

arteam committed:
Short version for starting a hot data node

arteam committed:
Fix the high watermark test name

arteam committed:
Run spotless

arteam committed:
Custer info NEEDS to be refreshed

arteam committed:
Move setBlockOnceOnReadSnapshotInfoIfAlreadyBlocked

arteam committed:
Check that indices get mounted and the extra one doesn't

arteam committed:
Assert that the index without enough memory doesn't get allocated

arteam committed:
Move getting mock repository

arteam committed:
Refresh cluster service info only once

arteam committed:
Merge remote-tracking branch 'origin/master' into copy-expected-shard-size

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDecider.java:
```diff
@@ -126,7 +127,11 @@ public static long sizeOfRelocatingShards(
         // Where reserved space is unavailable (e.g. stats are out-of-sync) compute a conservative estimate for initialising shards
         for (ShardRouting routing : node.shardsWithState(ShardRoutingState.INITIALIZING)) {
-            if (routing.relocatingNodeId() == null) {
+            IndexMetadata indexMetadata = metadata != null ? metadata.index(routing.index()) : null;
```
I think we can use `getIndexSafe` as in `getExpectedShardSize` and avoid the null checking? I think we should always have `metadata` here.

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDecider.java:
```diff
@@ -126,7 +127,11 @@ public static long sizeOfRelocatingShards(
+                continue;
+            }
+            if (routing.relocatingNodeId() == null && indexMetadata.isSearchableSnapshot() == false) {
                 // in practice the only initializing-but-not-relocating shards with a nonzero expected shard size will be ones created
```
```suggestion
                // in practice the only initializing-but-not-relocating non-searchable-snapshot shards with a nonzero expected shard size will be ones created
```

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDecider.java:
```diff
@@ -135,13 +140,37 @@ public static long sizeOfRelocatingShards(
+
+        // Count the STARTED searchable snapshot shards which are unaccounted in the cluster
+        for (ShardRouting shard : node.shardsWithState(ShardRoutingState.STARTED)) {
+            IndexMetadata indexMetadata = metadata.index(shard.index());
```
Use `getIndexSafe` and avoid the null check here?

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDecider.java:
```diff
@@ -135,13 +140,37 @@ public static long sizeOfRelocatingShards(
+            if (indexMetadata == null) {
+                continue;
+            }
+            if (indexMetadata.isSearchableSnapshot() == false) {
+                continue;
+            }
+            if (reservedSpace.containsShardId(shard.shardId())) {
+                continue;
+            }
+            if (clusterInfo.getShardSize(shard) != null) {
+                continue;
             }
```
Can we make this a single if with `&&` conditions instead? It reads slightly verbose and backwards now. (I do know the one above is similar, but anyway I would prefer that for this new new block of code).

henningandersen commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDeciderUnitTests.java:
```diff
@@ -471,16 +473,95 @@ public void testShardSizeAndRelocatingSize() {
+
+        Index index = new Index("test", "1234");
+        String nodeId = "node1";
+        ShardRouting shard1 = ShardRoutingHelper.moveToStarted(
```
Can we instead create the 3 shards in a loop? Seems overly verbose spelled out here.

arteam replied:
Not sure I'm following. I believe shards don't get filtered by the node id. `RoutingNode` just index all the shards that are supplied in its constructor.

henningandersen commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDeciderUnitTests.java:
```diff
@@ -471,16 +473,95 @@ public void testShardSizeAndRelocatingSize() {
+                    PeerRecoverySource.INSTANCE,
+                    new UnassignedInfo(UnassignedInfo.Reason.INDEX_CREATED, "foo")
+                ),
+                nodeId,
```
Can we add one additional shard that is added to another node? Just for checking that it does not impact it.

henningandersen commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/allocation/SearchableSnapshotDiskThresholdIntegTests.java:
```diff
@@ -268,6 +271,97 @@ public void testHighWatermarkCanBeExceededOnColdOrFrozenNode() throws Exception
+
+        String prefix = "mounted-";
+        mountIndices(indicesToBeMounted.keySet(), prefix, repository, snapshotName, FULL_COPY);
+        mockRepository.unblock();
```
I am not sure I understand why this unblocks here. Should it not wait until after we mounted the `indexToSkip`? Otherwise the test might only be checking against started indices.

arteam replied:
It's a good point. I actually not sure whether we need to unblock the repository at all.

henningandersen commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/allocation/SearchableSnapshotDiskThresholdIntegTests.java:
```diff
@@ -268,6 +271,97 @@ public void testHighWatermarkCanBeExceededOnColdOrFrozenNode() throws Exception
         });
     }
 
+    public void testOvercommitInitializingSearchableSnapshotShards() throws Exception {
```
Can we use a similar name to the previous test, since they sort of test the same but in different flavors?

henningandersen replied:
Also, I wonder why the tests are structured differently. The previous one mounts the same indices twice under different names. This one picks one not to mount, mounts the rest and then mounts the one extra index. I do not think there is a reason for that difference and maybe we can share most of the code between the two tests if we align the tests?

arteam replied:
I've extracted some common code of the tests into methods:

* createIndices
* createRepository
* createSnapshot
* mountIndices

henningandersen reviewed (commented):
Did another round on this direction and left more comments.

arteam committed:
Update server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDecider.java

Co-authored-by: Henning Andersen <33268011+henningandersen@users.noreply.github.com>

arteam committed:
Reformat

arteam committed:
Make moveToStarted accept `expectedShardSize`

arteam committed:
Use getIndexSafe for acquiring indexMetadata

arteam committed:
Pass ShardRouting.UNAVAILABLE_EXPECTED_SHARD_SIZE to moveToStarted

arteam committed:
Collapse if statements for searchable snapshots

arteam committed:
Correctly initialized started searchable snapshots

arteam committed:
Collapse shards to a loop

arteam committed:
Remove unused import

arteam committed:
Remove unblocking

arteam committed:
Extract creating a repository to a method

arteam committed:
Simplify starting a master and a hot node

arteam committed:
Remove unnecessary refreshing

arteam committed:
Update the the test name to make it more consistent

arteam committed:
Merge remote-tracking branch 'origin/master' into copy-expected-shard-size

arteam committed:
Rename to `createRepository`

fcofdez commented:
@arteam I think something went wrong while the branch was updated, there're some unrelated changes, can you check please? Additionally, it's preferred to avoid force pushing in-review branches, as github gets confused sometimes. Thanks!

arteam commented:
Hey @fcofdez, sorry for the force push, it was a trivial change of the visibility modifier for the `createRepository` method. 
I don't see any unrelated changes, I had to fix some merge conflicts in `DiskThresholdDeciderUnitTests` when pulling the latest changes from master, though.

fcofdez commented:
I can see this and other unrelated changes https://github.com/elastic/elasticsearch/pull/86331/files#diff-db4b8f7b56ce1418ffd2304e9a50f41140917dd02b6c7bfd36cc780cc6de9dbc

arteam commented:
Thank you! You're right, they are result of auto-formatting for the merge commit. I'll fix the changes.

arteam committed:
Merge remote-tracking branch 'origin/master' into copy-expected-shard-size

arteam commented:
Sorry, I had do redo the merge and force push the new change, because the previous was impossible to repair with a revert commit :(

fcofdez commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/allocation/SearchableSnapshotDiskThresholdIntegTests.java:
```diff
@@ -268,6 +271,90 @@ public void testHighWatermarkCanBeExceededOnColdOrFrozenNode() throws Exception
+        assertBusy(() -> {
+            var state = client().admin().cluster().prepareState().setRoutingTable(true).get().getState();
+            assertThat(
+                state.routingTable()
```
I think it's clearer to get the `IndexRoutingTable` for that particular index using `RoutingTable#index` and then assert that the index hasn't started, wdyt?

arteam replied:
Good idea! Change it to `state.routingTable().index(prefix + indexToSkip)`

fcofdez commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/allocation/SearchableSnapshotDiskThresholdIntegTests.java:
```diff
@@ -268,6 +271,90 @@ public void testHighWatermarkCanBeExceededOnColdOrFrozenNode() throws Exception
+        assertBusy(() -> {
+            var state = client().admin().cluster().prepareState().setRoutingTable(true).get().getState();
+            assertThat(
+                state.routingTable()
```
I think it's clearer to get the `IndexRoutingTable` for that particular index using `RoutingTable#index` and then assert that the index hasn't started, wdyt?

fcofdez reviewed (commented):
Looks mostly good, I left some comments in the tests.

arteam committed:
Merge remote-tracking branch 'origin/master' into copy-expected-shard-size

arteam committed:
Fix RoutingNode constructor

arteam committed:
Make shards operations more readable

arteam committed:
Randomly add a non-searchable snapshot index

arteam committed:
Count INITIALIZED shards

arteam committed:
Simplify checking the index that needs to be skipped

arteam committed:
Revert "Count INITIALIZED shards"

This reverts commit 21da01659e11cafa521d5b10553c0daca3bbcb97.

arteam commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDeciderUnitTests.java:
```diff
@@ -467,16 +471,73 @@ public void testShardSizeAndRelocatingSize() {
+        for (int i = 1; i <= 3; i++) {
+            int expectedSize = 10 * i;
+            shards.add(
+                ShardRoutingHelper.moveToStarted(
```
Split it.

arteam commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDeciderUnitTests.java:
```diff
@@ -467,16 +471,73 @@ public void testShardSizeAndRelocatingSize() {
+
+        Index index = new Index("test", "1234");
+        String nodeId = "node1";
+        List<ShardRouting> shards = new ArrayList<>();
```
Added

arteam commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/allocation/SearchableSnapshotDiskThresholdIntegTests.java:
```diff
@@ -268,6 +271,90 @@ public void testHighWatermarkCanBeExceededOnColdOrFrozenNode() throws Exception
+                    .filter(s -> indicesToBeMounted.containsKey(s.shardId().getIndexName().replace(prefix, "")))
+                    .filter(s -> state.metadata().index(s.shardId().getIndex()).isSearchableSnapshot())
+                    .filter(s -> otherDataNodeId.equals(s.currentNodeId()))
+                    .filter(s -> s.state() == ShardRoutingState.STARTED)
```
Shards never reliably go a to a state where all of them are `INITIALIZED`. It's always some mixture of `INITIALIZED` and `STARTED` until all the shards go to `STARTED`. I'd love to add a check that can verify that a certain amount of shards are being initialized, but I have no idea how to control it (`setBlockOnceOnReadSnapshotInfoIfAlreadyBlocked` increases the amount of INITIALIZED shards, but doesn't specifically set how many).

fcofdez replied:
I see what's going on here, `setBlockOnceOnReadSnapshotInfoIfAlreadyBlocked` only blocks once, we should have a new blob store implementation that blocks all the snapshot info reads until we deem it necessary. That way all the searchable snapshot shards should be in `ShardRoutingState.INITIALIZING` until we unblock the execution. We can pair to solve this if you want 👍.

arteam replied:
I've created a custom `CustomMockRepository` that locks on `restoreShard` which prevents shards to go from `INITIALIZED` to `STARTED`. Now we can check that we can go overboard and trip the watermark with initializing shards taking space. It uses a static `CountDownLatch` which I don't like and am open to suggestions how to implement it better!

arteam committed:
Merge remote-tracking branch 'origin/master' into copy-expected-shard-size

arteam committed:
Configure mock blob store with a count down latch

arteam committed:
Add a latch for initializing snapshots

arteam committed:
Remove debug printing statements

arteam committed:
Unlock AFTER we can't mount extra index

arteam committed:
Restore interrupted status for the interrupted thread

arteam committed:
Merge remote-tracking branch 'origin/master' into copy-expected-shard-size

fcofdez commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/allocation/SearchableSnapshotDiskThresholdIntegTests.java:
```diff
@@ -268,10 +283,159 @@ public void testHighWatermarkCanBeExceededOnColdOrFrozenNode() throws Exception
+        Map<String, Long> indicesStoresSizes = sizeOfShardsStores("index-*");
+        assertAcked(client().admin().indices().prepareDelete("index-*"));
+
+        String otherDataNode = internalCluster().startNode(
```
nit: it might be clearer to name it `coldNodeName`

fcofdez replied:
nit: it might be clearer to name it `coldNodeName`

fcofdez commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/allocation/SearchableSnapshotDiskThresholdIntegTests.java:
```diff
@@ -268,10 +283,159 @@ public void testHighWatermarkCanBeExceededOnColdOrFrozenNode() throws Exception
+
+        assertBusy(() -> {
+            var state = client().admin().cluster().prepareState().setRoutingTable(true).get().getState();
+            assertThat(
```
nit: Since there's just one index that should be unassigned, I think it might be clearer to assert that the shard is `UNASSIGNED`?

fcofdez replied:
nit: Since there's just one index that should be unassigned, I think it might be clearer to assert that the shard is `UNASSIGNED`?

fcofdez commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/allocation/SearchableSnapshotDiskThresholdIntegTests.java:
```diff
@@ -268,10 +283,159 @@ public void testHighWatermarkCanBeExceededOnColdOrFrozenNode() throws Exception
+            super(metadata, environment, namedXContentRegistry, clusterService, bigArrays, recoverySettings);
+        }
+
+        private void unlock() {
```
nit: `unlockRestore()`?

fcofdez replied:
nit: `unlockRestore()`?

fcofdez reviewed (commented):
I left a couple of small comments but this looks good to me, thanks for the iterations @arteam !

arteam committed:
Merge remote-tracking branch 'origin/master' into copy-expected-shard-size

arteam committed:
Rename to coldNodeId

arteam committed:
unlock -> unlockRestore

arteam committed:
Check that the extra shard is unassigned

arteam commented:
@elasticmachine run elasticsearch-ci/docs-check

fcofdez commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/allocation/SearchableSnapshotDiskThresholdIntegTests.java:
```diff
@@ -268,10 +283,156 @@ public void testHighWatermarkCanBeExceededOnColdOrFrozenNode() throws Exception
         });
     }
 
+    public void testHighWatermarkCanBeExceededOnColdNode() throws Exception {
```
nit: `testHighWatermarkCanNotBeExceededWithInitializingSearchableSnapshots`?

fcofdez reviewed (approved):
LGTM 👍. Let's wait for @henningandersen review before merging the changes.

arteam committed:
Change the test name

testHighWatermarkCanBeExceededOnColdNode ->
testHighWatermarkCanNotBeExceededWithInitializingSearchableSnapshots

arteam committed:
Merge remote-tracking branch 'origin/master' into copy-expected-shard-size

arteam committed:
Add missed import

arteam commented:
@elasticmachine update branch

arteam committed:
Merge remote-tracking branch 'origin/master' into copy-expected-shard-size

henningandersen commented on build-tools-internal/version.properties:
```diff
@@ -19,7 +19,7 @@ ecsLogging        = 1.2.0
 jna               = 5.10.0
 
-netty             = 4.1.77.Final
+netty=4.1.77.Final
```
Can we revert this, looks unrelated?

henningandersen commented on distribution/src/config/log4j2.properties:
```diff
@@ -63,8 +63,8 @@ appender.deprecation_rolling.name = deprecation_rolling
 # Intentionally follows a different pattern to above
-appender.deprecation_rolling.layout.dataset = deprecation.elasticsearch
-appender.deprecation_rolling.filter.rate_limit.type = RateLimitingFilter
+appender.deprecation_rolling.layout.dataset=deprecation.elasticsearch
+appender.deprecation_rolling.filter.rate_limit.type=RateLimitingFilter
```
Can we revert this, looks unrelated?

henningandersen reviewed (commented):
I started on this review, but it looks like a lot of unrelated changes have happened, making it a bit hard to review. Can you please revert all the unrelated changes?

arteam commented:
Ah, I already fixed it yesterday, but I didn't push the fixed merge commit. And sorry for the force push, unfortunately it's not straightforward to revert a mangled merge commit.

arteam commented:
@elasticmachine run elasticsearch-ci/docs

henningandersen reviewed (commented):
This looks mostly good, left 2 comments to look into.

arteam committed:
Remove the metadata null check

arteam committed:
Keep started shards in RoutingNode

arteam committed:
Remove unused import

arteam committed:
docs: Update changelog

arteam committed:
Merge remote-tracking branch 'origin/master' into copy-expected-shard-size

arteam commented:
@elasticmachine update branch

arteam committed:
Merge remote-tracking branch 'origin/master' into copy-expected-shard-size

arteam commented:
@elasticmachine run elasticsearch-ci/part-2

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDecider.java:
```diff
@@ -125,21 +126,37 @@ public static long sizeOfRelocatingShards(
+        }
+
+        // Count the STARTED searchable snapshot shards which are unaccounted in the cluster
+        for (ShardRouting shard : node.started()) {
```
I am a bit afraid of the time complexity here. With 1000 started shards here, this turns a bit expensive and runs many times on every reroute.

I'd propose to cache this in `RoutingAllocation` such that we cache the amount of total size to add by `RoutingNode` but also keep a "modification counter" from the `RoutingNode` in the cache to detect that it changed. We could also more explicitly clear the cache upon changes but that seems like a more advanced solution.

arteam replied:
Hey @henningandersen! I created a static cache in `RoutingAllocation` based on `RoutingNode`. If `RoutingNode` changes, the cache will calculate a new value because `equals&hasCode` will be changed too. I also added a TTL for the cache, so we update the cache based on fresh `ClusterServiceInfo`

henningandersen reviewed (commented):
This looks mostly good, though I have a couple more asks here.

arteam committed:
Merge remote-tracking branch 'origin/master' into copy-expected-shard-size

arteam committed:
Use Map.of

arteam committed:
Use getIndexSafe

arteam committed:
Precalculate size of unnacountable searchable snapshots

arteam committed:
Merge remote-tracking branch 'origin/master' into copy-expected-shard-size

arteam committed:
Use ShardRouting as a key

arteam committed:
Merge remote-tracking branch 'origin/main' into copy-expected-shard-size

arteam committed:
Revert "Use ShardRouting as a key"

This reverts commit 0104f1dd5510ab2affadfab2dbf6126573dba80e.

arteam committed:
Store unaccountableSearchableSnapshotSizes statically

Recalculate if RoutingNode gets changed

arteam committed:
Clear unnacountable searchable snapshot cache

arteam committed:
Revert "Clear unnacountable searchable snapshot cache"

This reverts commit b365aff19eb5e8083743fa2f8f1e38f3048c2909.

arteam committed:
Make unaccountableSearchableSnapshotSizes an LRU cache

arteam committed:
Use a TTL cache

arteam committed:
Update comment

arteam committed:
Merge remote-tracking branch 'origin/main' into copy-expected-shard-size

arteam commented:
@elasticmachine update branch

arteam committed:
Merge remote-tracking branch 'origin/main' into copy-expected-shard-size

arteam committed:
Merge remote-tracking branch 'origin/main' into copy-expected-shard-size

arteam committed:
Merge remote-tracking branch 'origin/main' into copy-expected-shard-size

arteam commented:
@elasticmachine
run elasticsearch-ci/bwc
run elasticsearch-ci/docs-check

arteam commented:
@elasticmachine update branch

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/RoutingAllocation.java:
```diff
@@ -333,6 +346,33 @@ public void setHasPendingAsyncFetch() {
+                long totalSize = 0;
+                // Count the STARTED searchable snapshot shards which are unaccounted in the cluster
+                for (ShardRouting shard : node.started()) {
+                    if (clusterInfo == null) {
+                        continue;
+                    }
```
Should this not go outside the loop - or even outside the `computeIfAbsent`?

arteam replied:
Good catch. I've moved it

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/RoutingAllocation.java:
```diff
@@ -333,6 +346,33 @@ public void setHasPendingAsyncFetch() {
+                        node.nodeId(),
+                        usage != null ? usage.getPath() : ""
+                    );
+                    if (clusterState.metadata().getIndexSafe(shard.index()).isSearchableSnapshot()
```
I wonder if we can check `shard.getExpectedShardSize() > 0` first, then check for `isSearchableSnapshot`, just to optimize a bit. Also, I'd like to defer looking up the `usage` and `reservedSpace` until we have determined that it is a searchable snapshot.

arteam replied:
Awesome! I've added the `shard.getExpectedShardSize() > 0` check and removed `Math.max(shard.getExpecredShardSize, 0) call

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/RoutingAllocation.java:
```diff
@@ -333,6 +346,33 @@ public void setHasPendingAsyncFetch() {
+                }
+                return totalSize;
+            });
+        } catch (ExecutionException ignore) {
```
Once we make `unaccountableSearchableSnapshotSizes` non-static, we can use a `Map` instead and I think this exception is then no longer needed. I'd prefer not to swallow exceptions.

arteam replied:
True that!

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/RoutingAllocation.java:
```diff
@@ -77,6 +83,13 @@ public class RoutingAllocation {
     private final DesiredNodes desiredNodes;
 
+    private static final TimeValue CACHE_TTL = TimeValue.timeValueMinutes(1);
+    private static final Cache<RoutingNode, Long> unaccountableSearchableSnapshotSizes = CacheBuilder.<RoutingNode, Long>builder()
```
I'd like to call the variable `unaccountedSearchableSnapshotSizes` instead. Also, please add javadoc describing what it means (searchable snapshots not yet registered in `ClusterInfo`).

arteam replied:
That indeed looks better! I've renamed it the field and the getter to `unaccountedSearchableSnapshotSizes` and added a comment to it.

henningandersen reviewed (commented):
A few more comments I'd like to see addressed.

arteam committed:
Merge remote-tracking branch 'origin/main' into copy-expected-shard-size

arteam committed:
Remove assertion UNAVAILABLE_EXPECTED_SHARD_SIZE hapens only for relocating and initializing shards

arteam commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/RoutingAllocation.java:
```diff
@@ -73,6 +79,13 @@ public class RoutingAllocation {
     private final Map<String, SingleNodeShutdownMetadata> nodeReplacementTargets;
 
+    private static final TimeValue CACHE_TTL = TimeValue.timeValueMinutes(1);
+    private static final Cache<RoutingNode, Long> unaccountableSearchableSnapshotSizes = CacheBuilder.<RoutingNode, Long>builder()
```
Yes, this was one the earlier designs, but then I went with the `static` approach

arteam committed:
Return to non-static unaccountableSearchableSnapshotSizes

arteam committed:
Move clusterInfo nullable check outside of the loop

arteam committed:
Check expectedShardSize is non-zero before checking that's is a searchable snapshot

arteam committed:
unnacoutable->unaccounted

fcofdez commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/RoutingAllocation.java:
```diff
@@ -119,6 +124,28 @@ public RoutingAllocation(
         }
         this.nodeReplacementTargets = Map.copyOf(targetNameToShutdown);
         this.desiredNodes = DesiredNodes.latestFromClusterState(clusterState);
+        unaccountedSearchableSnapshotSizes = new HashMap<>();
```
nit: can we assign the map at the end and make it unmodifiable?

fcofdez commented on server/src/main/java/org/elasticsearch/cluster/routing/RoutingNode.java:
```diff
@@ -67,7 +70,11 @@ private RoutingNode(RoutingNode original) {
         this.initializingShards = new LinkedHashSet<>(original.initializingShards);
+        this.startedShards = new LinkedHashSet<>(original.startedShards);
         this.shardsByIndex = Maps.copyOf(original.shardsByIndex, HashSet::new);
+        for (Map.Entry<Index, Set<ShardRouting>> entry : original.shardsByIndex.entrySet()) {
```
is this necessary? isn't this map copied in the previous line?

arteam replied:
Thank you for spotting this! That's a merge artifact from one the refactoring that happened during the development of the PR. I've removed it.

fcofdez commented on server/src/main/java/org/elasticsearch/cluster/routing/RoutingNode.java:
```diff
@@ -43,6 +43,8 @@ public class RoutingNode implements Iterable<ShardRouting> {
 
     private final LinkedHashSet<ShardRouting> relocatingShards;
 
+    private final LinkedHashSet<ShardRouting> startedShards;
```
can we add some tests to check that the started shards are handled correctly in this class?

arteam replied:
Good idea, I've added a test for checking that the `started` method accounts for added and removed nodes.

fcofdez commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitor.java:
```diff
@@ -423,13 +423,14 @@ public void onNewInfo(ClusterInfo info) {
     // exposed for tests to override
     long sizeOfRelocatingShards(RoutingNode routingNode, DiskUsage diskUsage, ClusterInfo info, ClusterState reroutedClusterState) {
-        return DiskThresholdDecider.sizeOfRelocatingShards(
+        return DiskThresholdDecider.sizeOfUnaccountedShards(
```
I'm not sure if it is correct to use 0L here? this might mess up the disk thresholds monitoring?

arteam replied:
Don't have access to `RoutingAllocation` here, so I'm not sure what's a better approach here.

fcofdez commented on server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java:
```diff
@@ -485,13 +482,14 @@ public ShardRouting reinitializeReplicaShard() {
      * <code>INITIALIZING</code> or <code>RELOCATING</code>. Any relocation will be
```
In `public ShardRouting(ShardId shardId, StreamInput in)` we're serializing/de-serializing the shard size only when the state is `state == ShardRoutingState.UNASSIGNED || state == ShardRoutingState.INITIALIZING` I think that's not consistent with the current behaviour?

arteam replied:
Good point! I've changed the serialization logic to write `expectedShardSize` for `STARTED` shards starting from `8.5.0`

fcofdez commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java:
```diff
@@ -571,8 +571,10 @@ private static void applyStartedShards(RoutingAllocation routingAllocation, List
                     + routingNodes.getByAllocationId(startedShard.shardId(), startedShard.allocationId().getId());
-
-            routingNodes.startShard(logger, startedShard, routingAllocation.changes());
+            long expectedShardSize = routingAllocation.metadata().getIndexSafe(startedShard.index()).isSearchableSnapshot()
```
I think this is always `ShardRouting.UNAVAILABLE_EXPECTED_SHARD_SIZE`? I cannot find where we populate this with the value from `org.elasticsearch.snapshots.SnapshotShardSizeInfo#getShardSize(org.elasticsearch.cluster.routing.ShardRouting, long)`. Am I missing something?

arteam replied:
It gets set in `ShardRouting#moveToStarted`

fcofdez replied:
I see it know, thanks for clarifying!

fcofdez reviewed (commented):
I left a few comments, the most important one is a question about when do we populate the shard expected size with the searchable snapshot size, I couldn't find where we do that, but maybe I'm just missing it. Thanks for iterating into this.

arteam committed:
Merge remote-tracking branch 'origin/main' into copy-expected-shard-size

arteam committed:
Merge remote-tracking branch 'origin/main' into copy-expected-shard-size

arteam committed:
Remove redundant copying of shardsByIndex

arteam committed:
Write expectedShardSize for STARTED shards to the wire

arteam committed:
Make sure unaccountedSearchableSnapshotSizes is unmodifiable

arteam committed:
Add a test for returning started shards

arteam committed:
Also check that relocating shards don't affect started shards

arteam commented:
@elasticmachine update branch

arteam committed:
Merge remote-tracking branch 'origin/main' into copy-expected-shard-size

fcofdez commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/RoutingAllocation.java:
```diff
@@ -119,6 +125,29 @@ public RoutingAllocation(
         }
         this.nodeReplacementTargets = Map.copyOf(targetNameToShutdown);
         this.desiredNodes = DesiredNodes.latestFromClusterState(clusterState);
+        Map<String, Long> unaccountedSearchableSnapshotSizes = new HashMap<>();
```
nit: Can we move this into a static method?

fcofdez commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDeciderUnitTests.java:
```diff
@@ -472,22 +477,113 @@ public void testShardSizeAndRelocatingSize() {
+            int expectedSize = 10 * i;
+            shards.add(createShard(index, nodeId, i - 1, expectedSize));
+            // randomly add shard for non-searchable snapshot index
+            if (randomBoolean()) {
```
isn't it possible to add more shards than expected for `another_index` here?

fcofdez commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDeciderUnitTests.java:
```diff
@@ -472,22 +477,113 @@ public void testShardSizeAndRelocatingSize() {
+            null,
+            0
+        );
+        long sizeOfRelocatingShards = sizeOfUnaccountedShards(
```
can we add some initializing searchable snapshot shards randomly here to ensure that those are accounted as well?

fcofdez reviewed (approved):
This LGTM, I left a couple of minor nits. Maybe we want to get @henningandersen +1 too here? Thanks for all the iterations!

arteam committed:
Merge remote-tracking branch 'origin/main' into copy-expected-shard-size

arteam committed:
Extract unaccountedSearchableSnapshotSizes to a static method

arteam committed:
Add more searchable snapshots on another index

arteam committed:
Add initializing shards to the mix

henningandersen reviewed (approved):
LGTM.

arteam committed:
Merge remote-tracking branch 'origin/main' into copy-expected-shard-size

arteam committed:
Move disk usage outsife of the loop

arteam commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/RoutingAllocation.java:
```diff
@@ -119,6 +125,33 @@ public RoutingAllocation(
+            for (RoutingNode node : clusterState.getRoutingNodes()) {
+                long totalSize = 0;
+                for (ShardRouting shard : node.started()) {
+                    DiskUsage usage = clusterInfo.getNodeMostAvailableDiskUsages().get(node.nodeId());
+                    ClusterInfo.ReservedSpace reservedSpace = clusterInfo.getReservedSpace(
+                        node.nodeId(),
+                        usage != null ? usage.getPath() : ""
+                    );
```
Good catch! It ended up in the inner loop as an artifact of one the review comments.

arteam committed:
Add a comment about what we return in unaccounted searchable snapshot sizes

arteam commented:
Thank you both!
